### PR TITLE
DHFPROD-3465: generating explorer options file on upgrade and when not present

### PR DESCRIPTION
**Pull request by rahulvudutala**

### CODE REVIEW

File: ml-data-hub-plugin/src/main/groovy/com/marklogic/gradle/task/GenerateExplorerQueryOptions.groovy [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+import com.marklogic.hub.impl.EntityManagerImpl
+import org.gradle.api.tasks.TaskAction
+
+class GenerateExplorerQueryOptions extends HubTask {
+
+    @TaskAction
+    void generateExplorerQueryOptions() {
+        EntityManagerImpl entityManager = (EntityManagerImpl) getEntityManager()
+        entityManager.generateExplorerQueryOptions()
```
[the remark this answers is not in the document]
Comment (rahulvudutala) [reply ml-data-hub-plugin/src/main/groovy/com/marklogic/gradle/task/GenerateExplorerQueryOptions.groovy]: @rjrudin I was about to ask that question. Should I add ```project.task("hubGenerateExplorerOptions", group: scaffoldGroup, type: GenerateExplorerQueryOptions,description: "Generates the query options files required for Explorer application").finalizedBy(["hubDeployUserModules"])``` task in DatahubPlugin.groovy class

File: marklogic-data-hub/src/main/java/com/marklogic/hub/impl/EntityManagerImpl.java [hunk 2]
```diff
@@ -99,6 +99,27 @@ public boolean saveQueryOptions() {
...
+    public void generateExplorerQueryOptions() {
+        QueryOptionsGenerator generator = new QueryOptionsGenerator(hubConfig.newStagingClient());
+        try {
+            Path dir = hubProject.getEntityConfigDir();
+            if (!dir.toFile().exists()) {
+                dir.toFile().mkdirs();
+            }
+            List<JsonNode> entities = getAllEntities();
+            File expStagingFile = Paths.get(dir.toString(), HubConfig.EXP_STAGING_ENTITY_QUERY_OPTIONS_FILE).toFile();
+            File expFinalFile = Paths.get(dir.toString(), HubConfig.EXP_FINAL_ENTITY_QUERY_OPTIONS_FILE).toFile();
```
[the remark this answers is not in the document]
Comment (rahulvudutala) [reply marklogic-data-hub/src/main/java/com/marklogic/hub/impl/EntityManagerImpl.java]: Addressed

File: ml-data-hub-plugin/src/main/groovy/com/marklogic/gradle/task/GenerateExplorerQueryOptions.groovy (hunk 1 above)
[the remark this answers is not in the document]
Comment (rahulvudutala) [reply ml-data-hub-plugin/src/main/groovy/com/marklogic/gradle/task/GenerateExplorerQueryOptions.groovy]: Added finalizedBy to hubGenerateExplorerOptions task

File: ml-data-hub-plugin/src/main/groovy/com/marklogic/gradle/DataHubPlugin.groovy [hunk 3]
```diff
@@ -108,6 +111,8 @@ class DataHubPlugin implements Plugin<Project> {
         project.task("hubSaveIndexes", group: scaffoldGroup, type: SaveIndexes,
             description: "Saves the indexes defined in {entity-name}.entity.json file to staging and final entity config in src/main/entity-config/databases directory")
 
+        project.tasks.mlPostDeploy.getDependsOn().add("hubGenerateExplorerOptions")
```
Comment (rjrudin) [inline ml-data-hub-plugin/src/main/groovy/com/marklogic/gradle/DataHubPlugin.groovy]: Nice - though try this out in a project that doesn't have any dependsOn on mlPostDeploy, just to verify that there's no null-pointer here.

**Pull request merged**